src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -43,6 +45,153 @@ void Bootstrap(const MatType& dataset,
+  {
+    bootstrapDataset = dataset;
+    bootstrapLabels = labels;
+    if (UseWeights)
+      bootstrapWeights = weights;
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ */
+template<bool UseWeights, typename G = std::mt19937>
+class SequentialBootstrap
+{
+ public:
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
+                               const G& generator = G(std::random_device()())) :
+    SequentialBootstrap(indicatorMatrix, indicatorMatrix.n_rows, generator)
+  {
+  }
+
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
+                               arma::uword sampleCount,
+                               const G& generator = G(std::random_device()())):
+    indicatorMatrix(indicatorMatrix),
+    colIndices(arma::linspace<arma::uvec>(0, indicatorMatrix.n_cols - 1)),
+    sampleCount(sampleCount),
+    generator(generator)
+  {
+  }
+
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    if (labels.n_cols != indicatorMatrix.n_rows)
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+          "labels and indicatorMatrix n_rows differ!");
+
+    // observations are stored as columns and dimensions
+    // (number of features) as rows.
+
+    const arma::uvec phi(ComputeSamples());
+
+    bootstrapDataset = dataset.cols(phi);
+    bootstrapLabels = labels.cols(phi);
+    if (UseWeights)
+      bootstrapWeights = weights.rows(phi);
+  }
+
+  arma::uvec ComputeSamples() const
+  {
+    arma::uvec phi;
+
+    while (phi.size() < sampleCount) {
+      const auto prob(ComputeNextDrawProbabilities(phi, indicatorMatrix));
+      std::discrete_distribution dist(prob.begin(), prob.end());
+
+      phi = arma::join_cols(phi, arma::uvec{ colIndices[dist(generator)] });
+    }
+
+    return phi;
+  }
+
+  static arma::vec ComputeAverageUniqueness(const arma::mat& indicatorMatrix)
+  {
+    // sum of each column
+    arma::rowvec concurrency(arma::sum(indicatorMatrix, 0));
+    auto uniqueness(
+      indicatorMatrix.each_row()
+      / concurrency.clamp(1.0,
+        std::numeric_limits<arma::rowvec::elem_type>::max()));
+    const auto n(arma::sum(indicatorMatrix, 1)); // sum of each row
+
+    return arma::sum(uniqueness, 1) / n; // mean of each row
+  }
+
+  static arma::vec ComputeNextDrawProbabilities(const arma::uvec& phi,
+                                             const arma::mat& indicatorMatrix)
+  {
+    arma::vec avg(indicatorMatrix.n_rows);
+    arma::uvec rows(phi);
+
+    rows.insert_rows(rows.n_rows, 1);
+    for (arma::uword i(0); i < avg.size(); ++i) {
+      rows.back() = i;
+      avg[i] = ComputeAverageUniqueness(indicatorMatrix.rows(rows)).back();
```

Comment:
If you are only using the last element of `ComputeAverageUniqueness()`, couldn't you just return that element?

Reply:
I kept the current interface so that the functionality can be easier checked in the unit test.

Reply:
I see; I think things may change due to the other comments I left, but in this case here it would amount to a significant extra allocation just for the sake of testing.  I would rather either change the interface so there is no extra calculation and test that, or, do higher level tests for the sequential bootstrap.  (For instance: if there is only one interval, it should look like uniform sampling.  At the other extreme, if there are like 1000 intervals that cover the same set of points, then those should show up way more in the output (or is it way less?  I might have my math upside down)).

Reply:
@rcurtin , this should now be addressed. I added another parameter to control how many output values are generated. So for the unit tests it can generate all values and for the production code it will only generate a single value (the .back()).